### PR TITLE
added AsyncEventSource & AsyncValueSource<T>

### PR DESCRIPTION
**Pull request by eitanhs**

### CODE REVIEW

**Review by realgpasternak — DISMISSED** [review]

**Review by realgpasternak — APPROVED** [review]

**Review by kpahima — DISMISSED** [review]

**Comment by realgpasternak** [comment]
Do you want to support timeout and/or cancellation of the wait?

File: Bsii.Dotnet.Utils/AsyncValueSource.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    /// <summary>
+    /// Provides <typeparamref name="T"/> values to all asynchrnously awaiting consumers<br/>
+    /// Note: all awaiters will get a reference to the same values (in case that <typeparamref name="T"/> is reference type)
+    /// </summary>
+    public class AsyncValueSource<T> : IAsyncValueProvider<T>
+    {
+        private TaskCompletionSource<T> _tcs = new TaskCompletionSource<T>(
+            TaskCreationOptions.RunContinuationsAsynchronously);
+
+        public void Next(T value)
```
Comment (gravikovich) [inline Bsii.Dotnet.Utils/AsyncValueSource.cs:14]: Maybe `SetNext` is a better name?

File: Bsii.Dotnet.Utils/AsyncValueSource.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+            var captured = _tcs;
+            _tcs = new TaskCompletionSource<T>(
+                TaskCreationOptions.RunContinuationsAsynchronously);
+            captured.SetResult(value);
+        }
+
+        public Task<T> GetNextAsync()
+        {
+            return _tcs.Task;
+        }
```
Comment (gravikovich) [inline Bsii.Dotnet.Utils/AsyncValueSource.cs:25]: why not
```suggestion
        public Task<T> GetNextAsync() => _tcs.Task;
```

File: Bsii.Dotnet.Utils/IAsyncValueProvider.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Threading.Tasks;
+
+namespace Bsii.Dotnet.Utils
+{
+    public interface IAsyncValueProvider<T>
```
Comment (gravikovich) [inline Bsii.Dotnet.Utils/IAsyncValueProvider.cs:5]: How about writing a test?

**Review by kpahima — APPROVED** [review]

**Review by realgpasternak — APPROVED** [review]

**Review by nraban — APPROVED** [review]

**Pull request merged**